src/_pytest/mark/structures.py

```diff
@@ -349,8 +349,9 @@ def __call__(self, *args: object, **kwargs: object):
         if args and not kwargs:
             func = args[0]
             is_class = inspect.isclass(func)
-            if len(args) == 1 and (istestfunc(func) or is_class):
```

Comment:
The mark should be stored on the actual test function. So in the case it is wrapped, we should store on `marking_func`

Reply:
That does look clearer. The one thing that bothers me is that this logic would need to be duplicated for `istestfunc` and `store_mark` because they both should use the actual func

Reply:
Hmm you are right, it was indeed better the way you did it before -- I missed that the unwrapping is also needed in `istestfunc`. Apologies for the back and forth...

So:
- Let's move the unwrapping back to the decorator
- I'd keep the explicit `if isinstance(obj, (staticmethod, classmethod)):` check, it's clearer and avoids unwrapping unexpected things for now.
- I'd change the name `marking_func` to `unwrapped_func` or such. I think by "marking func" you meant to say, "[we are] marking [this] func[tion]", but I initially read it as "[this is a] marking func[tion]", which is the opposite. So I suggest we avoid the confusion.